### PR TITLE
WIP: JAPI-372 Filepos fetch support

### PR DESCRIPTION
**Pull request by jpmcmu**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [ ] I have created a corresponding JIRA ticket for this submission
- [ ] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [ ] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRandomAccessFileReader.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            };
+
+            fetchRequestThread = new Thread(fetchRequestTask);
+            fetchRequestThread.start();
+        }
+
+        Object rslt = null;
+        try
+        {
+            rslt = this.binaryRecordReader.getNext();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRandomAccessFileReader.java:173]: does this need to be synchronized w/ the fetchreqThread?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRandomAccessFileReader.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+     * Reports summary of messages generated during read operation.
+     *
+     */
+    public void report()
+    {
+        if (getRemoteReadMessageCount() > 0)
+        {
+            log.warn("DataPartition '" + this.dataPartition + "' read operation messages:\n");
+            log.warn(getRemoteReadMessages());
+        }
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HpccRandomAccessFileReader.java:217]: is it worth reporting the case where no messages were read?

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java [hunk 3]
```diff
@@ -301,6 +301,60 @@ String createIndexOnDataset(String datasetName, FieldDef recordDef) throws Excep
...
+
+                Long[] recOffsets = {20L};
+                fileReader.addRecordRequests(Arrays.asList(recOffsets));
+
+                while (fileReader.hasNext())
+                {
+                    HPCCRecord record = fileReader.next();
+                    if (record == null)
+                    {
+                        System.out.println("Error: failed to read record.");
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java:338]: explicitly fail this test here

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java [hunk 4]
```diff
@@ -301,6 +301,60 @@ String createIndexOnDataset(String datasetName, FieldDef recordDef) throws Excep
...
+                    }
+
+                    long expectedKeyValue = 3 + 4 * i;
+                    Long keyValue = (Long) record.getField(0);
+                    if (keyValue == expectedKeyValue)
+                    {
+                        System.out.println("Error: key values did not match.");
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java:345]: this is the success case right? the message seems wrong.

File: dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java [hunk 5]
```diff
@@ -301,6 +301,60 @@ String createIndexOnDataset(String datasetName, FieldDef recordDef) throws Excep
...
+                    DataPartition fp = fileParts[(int)i];
+                    HPCCRecordBuilder recordBuilder = new HPCCRecordBuilder(file.getProjectedRecordDefinition());
+                    fileReader = new HpccRandomAccessFileReader<HPCCRecord>(fp, originalRD, recordBuilder,-1);
+                }
+                catch (Exception e)
+                {
+                    System.out.println("Error: " + e.getMessage());
+                }
+
+                Long[] recOffsets = {20L};
```
Comment (rpastrana) [inline dfsclient/src/test/java/org/hpccsystems/dfs/client/DFSIndexTest.java:330]: is this the 20th record, or the record at the 20th byte?

**Review by rpastrana — COMMENTED** [review]
@jpmcmu looks good, I know it's still in draft but pointed out a few things we might need to address.

**Comment by jpmcmu** [comment]
Closing, see updated PR: https://github.com/hpcc-systems/hpcc4j/pull/469

**Pull request closed without merging**